File: apps/tauri/src/lib/vpn/app-protocol-support.test.ts
import { describe, expect, it } from 'vitest'
import {
  isAppConnectable,
  isRelayCompatible,
  normalizeProtocol,
  unsupportedReason,
} from './app-protocol-support'
import type { NodeItem } from '@/api/client'

function node(
  protocol: string,
  accessMode: string | undefined = 'direct',
  tlsMode?: string,
): NodeItem {
  return {
    id: 1,
    name: 'test',
    region: 'cn',
    status: 'online',
    protocol,
    access_mode: accessMode,
    tls_mode: tlsMode,
  }
}

describe('app-protocol-support relay', () => {
  it('allows all sing-box protocols via relay (not only VLESS+Reality)', () => {
    for (const protocol of ['vless', 'vmess', 'trojan', 'shadowsocks', 'ss', 'hysteria2', 'hy2']) {
      const n = node(protocol, 'relay', 'tls')
      expect(isAppConnectable(n), protocol).toBe(true)
      expect(isRelayCompatible(n), protocol).toBe(true)
      expect(unsupportedReason(n), protocol).toBeNull()
    }
  })

  it('allows relay VLESS without Reality', () => {
    expect(isAppConnectable(node('vless', 'relay', 'tls'))).toBe(true)
  })

  it('rejects OpenVPN/WireGuard on relay', () => {
    for (const protocol of ['openvpn', 'wireguard']) {
      const n = node(protocol, 'relay')
      expect(isAppConnectable(n)).toBe(false)
      expect(isRelayCompatible(n)).toBe(false)
      // 原生栈优先提示官方客户端（与直连一致）
      expect(unsupportedReason(n) ?? '').toMatch(/官方客户端|OpenVPN|WireGuard/)
    }
  })

  it('rejects direct OpenVPN in App', () => {
    const n = node('openvpn', 'direct')
    expect(isAppConnectable(n)).toBe(false)
    expect(unsupportedReason(n)).toContain('OpenVPN')
  })

  it('normalizes protocol aliases', () => {
    expect(normalizeProtocol('hy2')).toBe('hysteria2')
    expect(normalizeProtocol('ss')).toBe('shadowsocks')
    expect(normalizeProtocol('')).toBe('vmess')
  })
})

File: apps/android/app/src/test/java/com/vpn/member/vpn/AppProtocolSupportTest.kt
package com.vpn.member.vpn

import com.vpn.member.data.api.NodeItem
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.assertEquals
import org.junit.Test

/** 经中转：自研 App 支持全部 sing-box 族；OpenVPN/WireGuard 仍不可连。 */
class AppProtocolSupportTest {
    private fun node(
        protocol: String,
        accessMode: String? = "direct",
        tlsMode: String? = null,
        status: String = "online",
    ) = NodeItem(
        id = 1L,
        name = "test",
        region = "cn",
        status = status,
        protocol = protocol,
        access_mode = accessMode,
        tls_mode = tlsMode,
    )

    @Test
    fun relaySingBoxProtocolsAreConnectable() {
        for (protocol in listOf("vless", "vmess", "trojan", "shadowsocks", "ss", "hysteria2", "hy2")) {
            val n = node(protocol, accessMode = "relay", tlsMode = "tls")
            assertTrue(protocol, AppProtocolSupport.isAppConnectable(n))
            assertTrue(protocol, AppProtocolSupport.isRelayCompatible(n))
            assertNull(protocol, AppProtocolSupport.unsupportedReason(n))
        }
    }

    @Test
    fun relayVlessWithoutRealityStillConnectable() {
        val n = node("vless", accessMode = "relay", tlsMode = "tls")
        assertTrue(AppProtocolSupport.isAppConnectable(n))
    }

    @Test
    fun relayOpenVpnAndWireGuardNotConnectable() {
        for (protocol in listOf("openvpn", "wireguard")) {
            val n = node(protocol, accessMode = "relay")
            assertFalse(protocol, AppProtocolSupport.isAppConnectable(n))
            assertFalse(protocol, AppProtocolSupport.isRelayCompatible(n))
            val reason = AppProtocolSupport.unsupportedReason(n)
            assertTrue(protocol, reason != null && reason.contains("官方客户端"))
        }
    }

    @Test
    fun directOpenVpnNotConnectable() {
        val n = node("openvpn", accessMode = "direct")
        assertFalse(AppProtocolSupport.isAppConnectable(n))
        assertEquals(
            "需使用 OpenVPN 官方客户端，自研 App 不支持",
            AppProtocolSupport.unsupportedReason(n),
        )
    }

    @Test
    fun directTrojanConnectable() {
        assertTrue(AppProtocolSupport.isAppConnectable(node("trojan", accessMode = "direct")))
    }

    @Test
    fun normalizeAliases() {
        assertEquals("hysteria2", AppProtocolSupport.normalizeProtocol("hy2"))
        assertEquals("shadowsocks", AppProtocolSupport.normalizeProtocol("ss"))
        assertEquals("vmess", AppProtocolSupport.normalizeProtocol(""))
    }
}
